Source/WebCore/platform/graphics/ComplexTextController.cpp

```diff
@@ -802,10 +802,16 @@ ComplexTextController::ComplexTextRun::ComplexTextRun(const Font& font, const UC
         auto currentIndex = r;
         char32_t character;
         U16_NEXT(m_characters, r, m_stringLength, character);
-        // https://drafts.csswg.org/css-text-3/#white-space-processing
+        // https://www.w3.org/TR/css-text-3/#white-space-processing
+        // "Control characters (Unicode category Cc)—other than tabs (U+0009), line feeds (U+000A), carriage returns (U+000D) and sequences that form a segment break—must be rendered as a visible glyph"
+        // Also, we're omitting Null (U+0000) from this set because Chrome and Firefox do so and it's needed for compat. See https://github.com/w3c/csswg-drafts/pull/6983.
+        if (character == nullCharacter || character == tabCharacter || character == newlineCharacter || character == carriageReturn)
```

Comment:
I wonder if this shouldn't be part of ``` instead. You have mentioned that this is already fixed for WidthIterator, which also uses this function, so maybe embedding this logic into `isCharacterWhoseGlyphsShouldBeDeletedForTextRendering` would allow us to eliminate some code on WidthIterator also.

Reply:
Fixed. Could you take a look again?